library/data/recipe/src/main/java/org/quiltmc/qsl/recipe/api/brewing/AbstractBrewingRecipe.java

```diff
@@ -0,0 +1,267 @@
+ */
+public abstract class AbstractBrewingRecipe<T> implements Recipe<BrewingStandBlockEntity> {
+	public static final List<Ingredient> VALID_INGREDIENTS = new ArrayList<>();
+	protected final T input;
+	protected final Ingredient ingredient;
+	protected final T output;
+	protected final int fuel;
+	protected final int brewTime;
+	protected ItemStack ghostOutput;
+	protected final String group;
+	private final Identifier id;
+
+	public AbstractBrewingRecipe(Identifier id, String group, T input, Ingredient ingredient, T output, int fuel, int brewTime) {
+		this.id = id;
+		this.group = group;
+		this.input = input;
+		this.ingredient = ingredient;
+		VALID_INGREDIENTS.add(ingredient);
+		this.output = output;
+		this.ghostOutput = new ItemStack(Items.POTION);
+		this.fuel = fuel;
+		this.brewTime = brewTime;
+	}
+
+	@Override
+	public RecipeType<AbstractBrewingRecipe<?>> getType() {
+		return Recipes.BREWING;
+	}
+
+	@Override
+	public ItemStack craft(BrewingStandBlockEntity inventory) {
+		for (int i = 0; i < 3; i++) {
+			if (this.matches(i, inventory.getStack(i))) {
+				inventory.setStack(i, this.craft(i, inventory.getStack(i)));
+			}
+		}
+
+		return ItemStack.EMPTY;
+	}
+
+	/**
+	 * Transforms the input {@link ItemStack} to the output {@link ItemStack}.
+	 * <p>
+	 * The input is guaranteed to match the required input, as defined by {@link #matches(int, ItemStack)}.
+	 *
+	 * @param slot the index of the slot
+	 * @param input the {@link ItemStack} in the provided slot
+	 * @return the output {@link ItemStack}
+	 */
+	protected abstract ItemStack craft(int slot, ItemStack input);
+
+	@Override
+	public boolean matches(BrewingStandBlockEntity inventory, World world) {
+		ItemStack ingredient = inventory.getStack(3);
+		if (this.ingredient.test(ingredient)) {
+			for (int i = 0; i < 3; ++i) {
+				if (matches(i, inventory.getStack(i))) {
+					return true;
+				}
+			}
+		}
+		return false;
+	}
+
+	/**
+	 * Matches based on stacks in the potion slots.
+	 *
+	 * @param slot the index of the slot
+	 * @param input the {@link ItemStack} in the provided slot
+	 * @return {@code true} if the {@link ItemStack} is a valid {@link AbstractBrewingRecipe#input} for this recipe, or {@code false}
+	 */
+	public abstract boolean matches(int slot, ItemStack input);
+
+	/**
+	 * {@return how much fuel this recipe takes to craft}
+	 */
+	public int getFuelUse() {
+		return this.fuel;
+	}
+
+	/**
+	 * {@return how long this recipe takes to craft}
+	 */
+	public int getBrewTime() {
+		return this.brewTime;
+	}
+
+	@Override
+	public boolean fits(int width, int height) {
+		return true;
+	}
+
+	@Override
+	public DefaultedList<ItemStack> getRemainder(BrewingStandBlockEntity inventory) {
+		// TODO integrate with custom recipe remainders
+		return Recipe.super.getRemainder(inventory);
+	}
+
+	@Override
+	public boolean isIgnoredInRecipeBook() {
+		return true;
+	}
+
+	@Override
+	public ItemStack createIcon() {
+		return new ItemStack(Blocks.BREWING_STAND);
```

Comment:
This seems like an odd choice; shouldn't the icon be related to the output?

Reply:
Icon is related to the category, so the crafting recipe icons are the crafting table, etc.